src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+            logger.debug("Disabling tombstone compactions for TOKCS");
+        }
+        else
+        {
+            logger.debug("Enabling tombstone compactions for TOKCS");
+        }
+    }
+
+    @Override
+    @SuppressWarnings("resource") // transaction is closed by AbstractCompactionTask::execute
+    public AbstractCompactionTask getNextBackgroundTask(int gcBefore)
+    {
+        SSTablesForCompaction previousCandidate = null;
+        while (true)
+        {
+            SSTablesForCompaction candidate = getNextBackgroundSSTables(gcBefore);
+
+            if (candidate == SSTablesForCompaction.EMPTY)
+            {
+                return null;
+            }
+
+            // Already tried acquiring references without success. It means there is a race with
+            // the tracker but candidate SSTables were not yet replaced in the compaction strategy
+            // manager
+            if (candidate.equals(previousCandidate))
+            {
+                logger.warn(
+                        "Could not acquire references for compacting SSTables {} which is not a problem per se,"
+                                + "unless it happens frequently, in which case it must be reported. Will retry later.",
+                        candidate);
+                return null;
+            }
+
+            LifecycleTransaction modifier = cfs.getTracker().tryModify(candidate.sstables, OperationType.COMPACTION);
+            if (modifier != null)
+            {
+                logger.debug("{}: {}", tableName, candidate);
+                return buildCompactionTask(candidate, modifier, gcBefore);
+            }
+            previousCandidate = candidate;
+        }
+    }
+
+    private AbstractCompactionTask buildCompactionTask(SSTablesForCompaction sstables, LifecycleTransaction modifier, int gcBefore)
+    {
+        gcBefore = sstables.tombstoneMerge ? CompactionManager.NO_GC : gcBefore;
+        return new TimeOrderedKeyCompactionTask(cfs, modifier, gcBefore, options, sstables.splitSSTable, sstables.tombstoneMerge);
+    }
+
+    /**
+     * Tombstone sstables can be of 3 categories:
+     * [expired, nearExpiry, latest]. See the inline documentation.
+     * <p>
+     * There are 2 parameters that will affect the performance of the operations on tables with this compaction strategy.
+     * 1. Time overlap between sstables. Many files overlaping with each other -> High read latency.
+     * 2. Time range of a sstable. Large range -> large size. Likelihood of it overlapping with other sstables also increases when the range is large, which
+     * will lead to larger cost in compaction. Compaction will be most efficient if there are fewer files to compact at a time.
+     * <p>
+     * This compaction strategy aims to reduce the max overlap and the number of sstables.
+     *
+     * @param gcBefore
+     * @return
+     */
+    synchronized SSTablesForCompaction getNextBackgroundSSTables(final int gcBefore)
```

Comment:
Why are all these synchronized? Are multiple threads going to be running/getting compaction tasks? Is a compaction strategy expected to be threadsafe?

Reply:
probably for a safety. Other classes also have it. I havent seen it running in parallel.